pycqed/measurement/waveform_control/segment.py

```diff
@@ -157,32 +157,66 @@ def resolve_timing(self):
         if len(visited_pulses) == 0:
             raise ValueError('No pulse references to the segment start!')
 
+        ref_pulses_dict_all = deepcopy(ref_pulses_dict)
         # add remaining pulses
-        while len(ref_points) > 0:
-            new_ref_points = []
-            for (name, pulse) in ref_points:
+        while len(ref_pulses_dict) > 0:
+            ref_pulses_dict_new = {}
+            for name, pulse in ref_pulses_dict.items():
                 for p in pulses[name]:
+                    if isinstance(p.ref_pulse, list):
```

Comment:
hard for me to evaluate just by reading the code but it looks good to me, also I know it has been tested by Christoph.